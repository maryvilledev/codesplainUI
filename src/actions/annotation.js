export const OPEN_ANNOTATION_PANEL = 'OPEN_ANNOTATION_PANEL';
export const CLOSE_ANNOTATION_PANEL = 'CLOSE_ANNOTATION_PANEL';

export const openAnnotationPanel = (snippetInformation) => ({
  type: OPEN_ANNOTATION_PANEL,
  payload: snippetInformation,
});

export const closeAnnotationPanel = () => ({
  type: CLOSE_ANNOTATION_PANEL,
});
