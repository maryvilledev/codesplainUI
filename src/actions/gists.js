export const SET_GISTS = 'SET_GISTS';

export const setGists = gists => ({
  type: SET_GISTS,
  payload: gists,
});
