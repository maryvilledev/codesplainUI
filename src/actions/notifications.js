export const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
export const CLOSE_NOTIFICATION = 'CLOSE_NOTIFICATION';

export const addNotification = notification => ({
  type: ADD_NOTIFICATION,
  payload: notification,
});

export const closeNotification = () => ({
  type: CLOSE_NOTIFICATION,
});
