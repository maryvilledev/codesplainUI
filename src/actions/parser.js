export const LOAD_PARSER = 'LOAD_PARSER'

export const loadParser = (parserURL) => ({
  type: LOAD_PARSER,
  meta: {
    WebWorker: true,
  },
  payload: parserURL,
})
