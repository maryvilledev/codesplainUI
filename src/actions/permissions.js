export const SET_PERMISSIONS = 'SET_PERMISSIONS';

export const setPermissions = permissions => ({
  type: SET_PERMISSIONS,
  payload: permissions,
});
