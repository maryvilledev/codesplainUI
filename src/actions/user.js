import axios from 'axios';
import cookie from 'react-cookie';

import { makeSaveEndpointUrl } from '../util/requests';

const API_URL = process.env.REACT_APP_API_URL;

// Util func to check for 'NoSuchKey' responses from S3
const noSuchKey = data => (
  typeof data === 'string' && data.includes('NoSuchKey')
);

export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const UPDATE_USER_SNIPPETS_STARTED = 'UPDATE_USER_SNIPPETS_STARTED';
export const UPDATE_USER_SNIPPETS_SUCCEEDED = 'UPDATE_USER_SNIPPETS_SUCCEEDED';
export const UPDATE_USER_SNIPPETS_FAILED = 'UPDATE_USER_SNIPPETS_FAILED';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';
export const SAVE_USERNAME = 'SAVE_USERNAME';
export const SAVE_ACCESS_TOKEN = 'SAVE_ACCESS_TOKEN';
export const CLEAR_USER_CREDENTIALS = 'CLEAR_USER_CREDENTIALS';
export const RESTORE_USER_CREDENTIALS = 'RESTORE_USER_CREDENTIALS';

export const saveUsername = username => ({
  type: SAVE_USERNAME,
  payload: username,
});

export const saveAccessToken = accessToken => ({
  type: SAVE_ACCESS_TOKEN,
  payload: accessToken,
});

export const clearUserCredentials = () => ({
  type: CLEAR_USER_CREDENTIALS,
});

export const setUserSnippets = snippetMeta => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const updateUserSnippetsStarted = () => ({
  type: UPDATE_USER_SNIPPETS_STARTED,
});

export const updateUserSnippetsSucceeded = () => ({
  type: UPDATE_USER_SNIPPETS_SUCCEEDED,
});

export const updateUserSnippetsFailed = () => ({
  type: UPDATE_USER_SNIPPETS_FAILED,
});

export const updateUserSnippets = () => (dispatch) => {
  // Load requisite cookies, return Promise if they aren't present
  const token = cookie.load('token');
  const username = cookie.load('username');

  // Fetch the user's snippet meta data and save it
  const headers = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  dispatch(updateUserSnippetsStarted());
  return axios.get(makeSaveEndpointUrl(username), { headers })
    .then((res) => {
      // Jump to catch block if the user has no index.json file:
      if (noSuchKey(res.data)) {
        throw new Error(`index.json does not exist for ${username}!`);
      }
      dispatch(setUserSnippets(res.data));
      dispatch(updateUserSnippetsSucceeded());
    })
    .catch(() => {
      dispatch(updateUserSnippetsFailed());
    });
};

export const fetchAccessToken = authCode => () => {
  const reqUrl = `${API_URL}/auth`;
  const reqBody = { code: authCode };
  return axios({
    method: 'POST',
    url: reqUrl,
    data: reqBody,
  });
};

export const fetchUserInfo = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user',
    headers: reqHeaders,
  });
};

export const restoreUserCredentials = (token, username) => ({
  type: RESTORE_USER_CREDENTIALS,
  payload: {
    token,
    username,
  },
});
