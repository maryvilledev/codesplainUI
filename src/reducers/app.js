import _ from 'lodash';

import * as actions from '../actions/app';
import { generateFilters } from '../util/codemirror-utils';

export const initialState = {
  annotations: {},
  AST: {},
  filters: {},
  hasUnsavedChanges: false,
  snippetLanguage: 'python3',
  readOnly: false,
  snippet: '',
  snippetKey: '',
  snippetTitle: '',
};

const app = (state = initialState, action) => {
  switch (action.type) {
  case actions.RESET_STATE: {
    return {
      ...initialState,
    };
  }
  case actions.SET_SNIPPET_KEY: {
    return {
      ...state,
      snippetKey: action.payload,
    };
  }
  case actions.SET_SNIPPET_CONTENTS: {
    return {
      ...state,
      hasUnsavedChanges: true,
      snippet: action.payload,
    };
  }
  case actions.SET_RULE_FILTERS: {
    return {
      ...state,
      filters: action.payload,
      hasUnsavedChanges: true,
    };
  }
  case actions.TOGGLE_EDITING_STATE: {
    if (state.readOnly) {
      return state;
    }
    return {
      ...state,
      hasUnsavedChanges: true,
      readOnly: !state.readOnly,
    };
  }
  case actions.SELECT_ALL_FILTERS: {
    const filters = _.mapValues(state.filters, (filter) => {
      if (filter.selected) {
        return filter;
      }
      return { ...filter, selected: true };
    });
    return {
      ...state,
      filters,
    };
  }
  case actions.RESET_RULE_FILTERS: {
    const filters = _.mapValues(state.filters, (filter) => {
      if (filter.selected) {
        return { ...filter, selected: false };
      }
      return filter;
    });
    return {
      ...state,
      filters,
    };
  }
  case actions.SET_AST: {
    return {
      ...state,
      AST: action.payload,
    };
  }
  case actions.SAVE_ANNOTATION: {
    return {
      ...state,
      annotations: {
        ...state.annotations,
        [action.payload.lineNumber]: action.payload,
      },
      hasUnsavedChanges: true,
    };
  }
  case actions.RESTORE_STATE: {
    return {
      ...action.payload,
      hasUnsavedChanges: false,
    };
  }
  case actions.SET_SNIPPET_TITLE: {
    return {
      ...state,
      hasUnsavedChanges: true,
      snippetTitle: action.payload,
    };
  }
  case actions.SAVE_SUCCEEDED: {
    return {
      ...state,
      hasUnsavedChanges: false,
    };
  }
  case actions.SET_SNIPPET_LANGUAGE: {
    return {
      ...state,
      hasUnsavedChanges: true,
      snippetLanguage: action.payload,
    };
  }
  case actions.CLEAR_UNSAVED_CHANGES: {
    return {
      ...state,
      hasUnsavedChanges: false,
    };
  }
  case actions.PARSE_SNIPPET: {
    if (action.meta) {
      return state;
    }
    const {
        AST,
        ruleCounts,
      } = action.payload;
    const filters = generateFilters(state.filters, ruleCounts);
    return {
      ...state,
      AST,
      filters,
    };
  }
  default: {
    return state;
  }
  }
};

export default app;
