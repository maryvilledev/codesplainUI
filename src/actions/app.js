import axios from 'axios';

import { addNotification, closeNotification } from './notifications';
import { makeSaveEndpointUrl } from '../util/requests';

export const RESET_STATE = 'RESET_STATE';
export const EDIT_ANNOTATION = 'EDIT_ANNOTATION';
export const PARSE_SNIPPET = 'PARSE_SNIPPET';
export const RESET_RULE_FILTERS = 'RESET_RULE_FILTERS';
export const RESTORE_STATE = 'RESTORE_STATE';
export const SAVE_ANNOTATION = 'SAVE_ANNOTATION';
export const SAVE_FAILED = 'SAVE_FAILED';
export const SAVE_STARTED = 'SAVE_STARTED';
export const SAVE_SUCCEEDED = 'SAVE_SUCCEEDED';
export const SELECT_ALL_FILTERS = 'SELECT_ALL_FILTERS';
export const SET_AST = 'SET_AST';
export const SET_RULE_FILTERS = 'SET_RULE_FILTERS';
export const SET_SNIPPET_CONTENTS = 'SET_SNIPPET_CONTENTS';
export const SET_SNIPPET_LANGUAGE = 'SET_SNIPPET_LANGUAGE';
export const SET_SNIPPET_TITLE = 'SET_SNIPPET_TITLE';
export const TOGGLE_EDITING_STATE = 'TOGGLE_EDITING_STATE';
export const SAVE_STATE = 'SAVE_STATE';
export const SAVE_STATE_STARTED = 'SAVE_STATE_STARTED';
export const SAVE_STATE_SUCCEEDED = 'SAVE_STATE_SUCCEEDED';
export const SAVE_STATE_FAILED = 'SAVE_STATE_FAILED';
export const SET_SNIPPET_KEY = 'SET_SNIPPET_KEY';

export const setSnippetKey = key => ({
  type: SET_SNIPPET_KEY,
  payload: key,
});

export const resetState = () => ({
  type: RESET_STATE,
});

export const setSnippetContents = snippet => ({
  type: SET_SNIPPET_CONTENTS,
  payload: snippet,
});

export const setRuleFilters = filters => ({
  type: SET_RULE_FILTERS,
  payload: filters,
});

export const selectAllFilters = () => ({
  type: SELECT_ALL_FILTERS,
});

export const resetFilters = () => ({
  type: RESET_RULE_FILTERS,
});

export const setAST = AST => ({
  type: SET_AST,
  payload: AST,
});

export const toggleEditState = () => ({
  type: TOGGLE_EDITING_STATE,
});

export const saveAnnotation = annotationData => ({
  type: SAVE_ANNOTATION,
  payload: annotationData,
});

export const restoreState = savedState => ({
  type: RESTORE_STATE,
  payload: savedState,
});

export const setSnippetTitle = snippetTitle => ({
  type: SET_SNIPPET_TITLE,
  payload: snippetTitle,
});

export const setSnippetLanguage = snippetLanguage => ({
  type: SET_SNIPPET_LANGUAGE,
  payload: snippetLanguage,
});

export const parseSnippet = snippet => ({
  type: PARSE_SNIPPET,
  meta: {
    WebWorker: true,
  },
  payload: {
    snippet,
  },
});

export const saveSucceeded = () => ({
  type: SAVE_SUCCEEDED,
});

export const saveNew = org => (dispatch, getState) => {
  // Load the token from cookie storage
  const {
    app: appState,
    user: { token },
  } = getState();
  // Construct the necessary request objects
  const reqBody = JSON.stringify(appState);
  const reqHeaders = {
    headers: {
      Authorization: token,
    },
  };
  dispatch(addNotification('Saving...'));
    // Save the new snippet
  return axios.post(makeSaveEndpointUrl(org), reqBody, reqHeaders)
      .then((res) => {
        // Remove the 'saving...' notification
        dispatch(closeNotification());
        // Give user feedback that saving succeeded
        dispatch(addNotification('Codesplaination Saved!'));
        // Clear the hasUnsavedChanges flag
        dispatch(saveSucceeded());
        // Return the key of the newly-saved snippet so that the browser
        // location can be updated
        return res.data.key;
      }, () => {
        // Give user feedback that saving failed
        dispatch(addNotification('Failed to save snippet; please try again'));
      });
};

export const saveExisting = () => (dispatch, getState) => {
    // Get the app state to save (and the snippet title to save to)
  const {
    app: appState,
    user: {
      token,
      username,
    },
  } = getState();
  const { snippetKey: key } = appState;

    // Construct the necessary request objects
  const reqBody = JSON.stringify(appState);
  const reqHeaders = {
    headers: {
      Authorization: token,
    },
  };
  dispatch(addNotification('Saving...'));
    // Update the snippet
  return axios.put(makeSaveEndpointUrl(username, key), reqBody, reqHeaders)
      .then(() => {
        // Remove the 'saving...' notification
        dispatch(closeNotification());
        // Give user feedback that saving succeeded
        dispatch(addNotification('Codesplaination Saved!'));
        // Clear the hasUnsavedChanges flag
        dispatch(saveSucceeded());
      }, () => {
        // Give user feedback that saving failed
        dispatch(addNotification('Failed to save snippet; please try again'));
      });
};

export const loadSnippet = (username, snippetKey) => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Authorization: token,
  };
  return axios({
    method: 'GET',
    url: makeSaveEndpointUrl(username, snippetKey),
    headers: reqHeaders,
  });
};
