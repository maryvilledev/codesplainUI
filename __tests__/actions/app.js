import moxios from 'moxios';
import cookie from 'react-cookie';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import * as actions from '../../src/actions/app';

describe('Actions: App', () => {
  describe('SET_SNIPPET_KEY', () => {
    it('creates correct action object', () => {
      const key = 'szechuan sauce';
      expect(actions.setSnippetKey(key)).toMatchSnapshot();
    });
  });
  describe('SET_SNIPPET_CONTENTS', () => {
    it('creates an action to set the snippet contents', () => {
      const snippet = 'Show me what you got';
      const expected = {
        type: actions.SET_SNIPPET_CONTENTS,
        payload: snippet,
      };
      expect(actions.setSnippetContents(snippet)).toEqual(expected);
    });
  });
  describe('SET_RULE_FILTERS', () => {
    it('creates an action to set the rule filters', () => {
      const filters = {
        for_stmt: {
          prettyTokenName: 'For Loops',
          count: 1,
          selected: false,
        },
        atom: {
          prettyTokenName: 'Atoms',
          count: 21,
          selected: true,
        },
      };
      const expected = {
        type: actions.SET_RULE_FILTERS,
        payload: filters,
      };
      expect(actions.setRuleFilters(filters)).toEqual(expected);
    });
  });
  describe('RESET_RULE_FILTERS', () => {
    it('matches snapshot', () => {
      expect(actions.resetFilters()).toMatchSnapshot();
    });
  });
  describe('SELECT_ALL_FILTERS', () => {
    it('matches snapshot', () => {
      expect(actions.selectAllFilters()).toMatchSnapshot();
    });
  });
  describe('RESET_STATE', () => {
    it('matches snapshot', () => {
      expect(actions.resetState()).toMatchSnapshot();
    });
  });
  describe('SET_AST', () => {
    it('creates an action to set the abstract syntax tree', () => {
      const AST = {
        type: 'file_input',
        begin: 0,
        end: 1,
        children: [
          'NEWLINE',
          'INDENT',
          'NEWLINE',
          'DEDENT',
        ],
      };
      const expected = {
        type: actions.SET_AST,
        payload: AST,
      };
      expect(actions.setAST(AST)).toEqual(expected);
    });
  });
  describe('TOGGLE_EDITING_STATE', () => {
    it('creates an action that toggles the editing state', () => {
      const expected = {
        type: actions.TOGGLE_EDITING_STATE,
      };
      expect(actions.toggleEditState()).toEqual(expected);
    });
  });
  describe('SAVE_ANNOTATION', () => {
    it('creates an action that adds an annotation', () => {
      const annotationData = {
        annotation: 'You pass butter.',
        lineNumber: 1,
        lineText: 'What is my purpose?',
      };
      const expected = {
        type: actions.SAVE_ANNOTATION,
        payload: annotationData,
      };
      expect(actions.saveAnnotation(annotationData)).toEqual(expected);
    });
  });
  describe('RESTORE_STATE', () => {
    it('creates an action that restores state', () => {
      const savedState = {
        snippet: '\t',
        snippetTitle: '',
        annotations: {},
        AST: {
          type: 'file_input',
          begin: 0,
          end: 1,
          children: [
            'NEWLINE',
            'INDENT',
            'NEWLINE',
            'DEDENT',
          ],
        },
        filters: {},
        readOnly: false,
      };
      const expected = {
        type: actions.RESTORE_STATE,
        payload: savedState,
      };
      expect(actions.restoreState(savedState)).toEqual(expected);
    });
  });
  describe('SET_SNIPPET_TITLE', () => {
    it('creates an action to set the snippet title', () => {
      const snippetTitle = 'Get Schwifty';
      const expected = {
        type: actions.SET_SNIPPET_TITLE,
        payload: snippetTitle,
      };
      expect(actions.setSnippetTitle(snippetTitle)).toEqual(expected);
    });
  });
  describe('SET_SNIPPET_LANGUAGE', () => {
    it('creates an action to set the snippet language', () => {
      const snippetLanguage = 'Gromflomite';
      const expected = {
        type: actions.SET_SNIPPET_LANGUAGE,
        payload: snippetLanguage,
      };
      expect(actions.setSnippetLanguage(snippetLanguage)).toEqual(expected);
    });
  });
  describe('PARSE_SNIPPET', () => {
    it('creates an action to parse the snippet with a web worker', () => {
      const snippet = 'Grasssss tastes bad';
      const expected = {
        type: actions.PARSE_SNIPPET,
        meta: {
          WebWorker: true,
        },
        payload: {
          snippet,
        },
      };
      expect(actions.parseSnippet(snippet)).toEqual(expected);
    });
  });
  describe('SAVE_STARTED', () => {
    it('creates the correct action object', () => {
      const expected = {
        type: actions.SAVE_STARTED,
      };
      expect(actions.saveStarted()).toEqual(expected);
    });
  });
  describe('SAVE_SUCCEEDED', () => {
    it('creates the correct action object', () => {
      const expected = {
        type: actions.SAVE_SUCCEEDED,
      };
      expect(actions.saveSucceeded()).toEqual(expected);
    });
  });
  describe('SAVE_FAILED', () => {
    it('creates the correct action object', () => {
      const expected = {
        type: actions.SAVE_FAILED,
      };
      expect(actions.saveFailed()).toEqual(expected);
    });
  });
  describe('async actions', () => {
    const middlewares = [thunk];
    const mockStore = configureMockStore(middlewares);

    beforeEach(() => {
      cookie.save('token', 'rick');
      cookie.save('username', 'morty');
      moxios.install();
    });
    afterEach(() => {
      cookie.remove('token');
      cookie.remove('username');
      moxios.uninstall();
    });
    describe('saveNew()', () => {
      it('dispatches SAVE_SUCCEEDED if saved', () => {
        const key = 'gazorpazorp';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
            response: { key, status: '200' },
          });
        });
        const expectedActions = [
          { type: actions.SAVE_STARTED },
          { type: actions.SAVE_SUCCEEDED },
        ];
        const store = mockStore({});
        return store.dispatch(actions.saveNew())
          .then((returnVal) => { // return of async actions
            expect(store.getActions()).toEqual(expectedActions);
            expect(returnVal).toEqual(key);
          });
      });
      it('dispatches SAVE_FAILED if save failed', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const expectedActions = [
          { type: actions.SAVE_STARTED },
          { type: actions.SAVE_FAILED },
        ];
        const store = mockStore({});
        return store.dispatch(actions.saveNew())
          .then(() => { // return of async actions
            expect(store.getActions()).toEqual(expectedActions);
          });
      });
    });
    describe('saveExisting()', () => {
      it('description', () => {
        it('dispatches SAVE_SUCCEEDED if saved', () => {
          moxios.wait(() => {
            const request = moxios.requests.mostRecent();
            request.respondWith({
              status: 200,
              response: { status: '200' },
            });
          });
          const expectedActions = [
            { type: actions.SAVE_STARTED },
            { type: actions.SAVE_SUCCEEDED },
          ];
          const store = mockStore({ snippetTitle: 'foo' });
          return store.dispatch(actions.saveExisting())
            .then(() => { // return of async actions
              expect(store.getActions()).toEqual(expectedActions);
            });
        });
        it('dispatches SAVE_FAILED if save failed', () => {
          moxios.wait(() => {
            const request = moxios.requests.mostRecent();
            request.respondWith({
              status: 400,
            });
          });
          const expectedActions = [
            { type: actions.SAVE_STARTED },
            { type: actions.SAVE_FAILED },
          ];
          const store = mockStore({});
          return store.dispatch(actions.saveExisting())
            .then(() => { // return of async actions
              expect(store.getActions()).toEqual(expectedActions);
            });
        });
      });
    });
  });
});
