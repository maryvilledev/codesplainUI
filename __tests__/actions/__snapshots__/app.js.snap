exports[`Actions: App RESET_RULE_FILTERS matches snapshot 1`] = `
Object {
  "type": "RESET_RULE_FILTERS",
}
`;

exports[`Actions: App RESET_STATE matches snapshot 1`] = `
Object {
  "type": "RESET_STATE",
}
`;

exports[`Actions: App SELECT_ALL_FILTERS matches snapshot 1`] = `
Object {
  "type": "SELECT_ALL_FILTERS",
}
`;

exports[`Actions: App SET_SNIPPET_KEY creates correct action object 1`] = `
Object {
  "payload": "szechuan sauce",
  "type": "SET_SNIPPET_KEY",
}
`;

exports[`Actions: App async actions saveNew() dispatches SAVE_FAILED if save failed 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "payload": "Failed to save snippet; please try again",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions saveNew() dispatches SAVE_SUCCEEDED if saved 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Codesplaination Saved!",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "SAVE_SUCCEEDED",
  },
]
`;
